File: 0098 - Validate Binary Search Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode], mi=-inf, ma=inf) -> bool:
        if not root:
            return True
        if not (mi < root.val < ma):
            return False
        return self.isValidBST(root.left, mi, root.val) and self.isValidBST(root.right, root.val, ma)

File: 0098 - Validate Binary Search Tree/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool solve(struct TreeNode *root, int **pre){
    if(!root) return true;
    if(!solve(root->left, pre)) return false;
    if(*pre && root->val<=**pre) return false;
    *pre = &(root->val);
    if(!solve(root->right, pre)) return false;
    return true;
}
bool isValidBST(struct TreeNode* root){
    int *pre = NULL;
    return solve(root, &pre);
}